### ISSUE TITLE
Question on the Latest Windows64 Release

### ISSUE THREAD

**Issue by Splee99** [lightvector/KataGo#109]
I downloaded the Katago 1.3.1 Windows 64 bit release and it seems to run OK, but the program seems to be in single thread mode, even I have a two core CPU. So the CPU loading is always around 50%. This is not the case for Leelaz program. It is similar to the earlier Katago release. Is there any setting that I should check to make sure the two CPU cores are activated?

**Comment by kennychenfs** [comment]
Do you set the config? The search threads might be what you're looking for.

**Comment by Splee99** [comment]
I have set this:
numSearchThreads = 21

**Comment by Splee99** [comment]
OK. I'm running the benchmark to find the best thread setting. Hopefully it can find a better one.

**Comment by lightvector** [comment]
Depending on your GPU, (a real GPU, or just an integrated graphics, or something else?) - KataGo could be spending a lot of its time waiting on the GPU to respond, and the CPU will have nothing to do while this is the case - so it might be 50% just by chance - using both CPUs on but the waiting makes it spend about half its time idling? 

Or if you see it literally pinned at exactly 50% all the time, regardless of threads, maybe you have something external to KataGo configured somehow weird. I never added any code to KataGo that would limit how many CPU cores used.

Yes, try running the benchmark.

**Comment by Splee99** [comment]
The benchmark indicated that I should use the thread setting of 6 instead of 21. However, I still get the CPU usage of only 50% to 60%. It's not related to the number of cores, just CPU usage. This is different than LZ, where I got obvious CPU usage increase when I set more threats using the "-t" option. Currently I set "-t 8" for LZ although the CPU only has two cores and the CPU usage is about 80%.

I'll try the opencl version of Katago.

**Comment by lightvector** [comment]
Okay, so clearly you're not pinned to 50%, since you said 50% to 60%. So there's probably nothing wrong, so please stop worrying about it. KataGo doesn't use all the CPU on your computer because your GPU can't keep up. There is no benefit to trying to make it use more CPU, because CPU is not the bottleneck.

**Comment by Splee99** [comment]
The main reason I'm worrying about this is because there is zero chance that Katago can beat ELF on my computer using equal thinking time. Then I'm confused when somebody says that Katago is stronger than ELF. So since the slower CPU is not the bottle neck, there must be something else and the GPU is certainly not running at its full potential.

**Comment by lightvector** [comment]
Why do you say there is zero chance? 
* When you say ELF, do you mean ELF within LZ? If so, can you post benchmark results for LZ, or at least how many playouts per second you observe, and benchmark results for KataGo (`katago.exe benchmark ...`) as well?
* Can you say what your GPU is? NVIDIA? AMD? Intel Graphics? What exact model and version?
* Can you post your GTP config for KataGo?
* Are you using CUDA or OpenCL for KataGo? If you've gotten both to work, are both slow for you?

With the latest released nets, KataGo should be at least 200 Elo stronger than ELFv2 at equal playouts, for numbers of playouts around high hundreds / low thousands. So presumably you're seeing a massive discrepancy between the number of playouts both can get in the same time? How big of a difference exactly?

**Comment by Splee99** [comment]
I just found that the Opencl version is much faster and Katago did manage to win one game with ELF (v2). However, it also lost several games and I will test further. Attached is the log from ELF and Katago, also included is the Katago config file.

[ELF-Katago-log.zip](https://github.com/lightvector/KataGo/files/4115651/ELF-Katago-log.zip)

Thank you for your attention.

**Comment by lightvector** [comment]
It's hard for me to tell if something is unusual about KataGo on your hardware or not. It's possible that the answer simply might be: On GTX 1050 (your graphics card), Leela Zero might just be optimized better for performance, whereas on other graphics cards that certain other users had been testing on, the difference might be less. Based on your logs, it seems that often LZ-ELF is getting 1.5x to 2x the playouts that KataGo is getting despite being the same sized network, so despite KataGo's network being stronger in raw terms, there's certainly a hardware/software performance gap of some sort.

I can offer a few things though:
* Based on your log, you're not using the latest KataGo net. The newer one should be a decent bit stronger than yours, see here: https://github.com/lightvector/KataGo/releases
* Given that you're finding the OpenCL version to be better, you might try the attached executable. This is a (very) recent unreleased branch that attempts to improve OpenCL speed. [katago-openclopt.zip](https://github.com/lightvector/KataGo/files/4115721/katago-openclopt.zip)
* nnCacheSizePowerOfTwo = 26 in your config seems very large. Do you actually have 100 GB of RAM (2^26 * 1.5 KB), and are you really ever searching for an amount of time where you want to cache 2^26 = ~64 million playouts worth of search in order use all of that RAM, even over the duration of an entire game?
* You might try retuning the number of search threads if the value you tuned was based on the CUDA benchmark.

**Comment by lightvector** [comment]
Also, it's probably not something stupid like this, but just in case: depending on how you're running LZ, you might if LZ is being set to ponder (i.e. to think on the opponent's time). You can explicitly provide `--noponder` if you aren't already, to make sure.

**Comment by Splee99** [comment]
It's already set to noponder. I will reduce the cache number. Also what is nnMutexPoolSizePowerOfTwo? Is my setting of 17 reasonable?

**Comment by lightvector** [comment]
Yes, your 17 is fine, even the default of 15 would be fine. Basically, 17 means that there are 2^17 = ~130K different mutexes such that threads will randomly pick different ones whenever they check for a cached neural net evaluation. When two threads collide, it costs a tiny tiny bit of performance briefly. Given that you're only using like 6 threads, having >100000 of these is already overkill.

The cache size itself doesn't matter much once it's a decent factor larger than how long you might search in a single position. The default is 19 to be friendlier on memory usage and because 2^19 = ~500K playouts is already well on the high end of how many playouts a casual user would usually run in a single search, except perhaps on top-of-the-line hardware.

The executable I linked is should be a drop-in replacement for v1.3.1's OpenCL version. I had posted it elsewhere as well, to try to get feedback on whether it is indeed faster - if so it's likely to go out in the next release. If you'd like to test it out as well and see, that would be useful feedback too. Same thing for the newer neural net - let me know if you encounter any problems - new net won't fix the performance issue, but might make up some of the gap purely due to improved strength, even if performance stays worse.

**Comment by Splee99** [comment]
OK. Thanks.

**Comment by Splee99** [comment]
The new Opencl version is indeed much better, about 40-50% faster from the benchmark comparison. Great work!

**Comment by lightvector** [comment]
Everything here seems handled? Closing. Feel free to reply back or to open a new issue for further questions.
